Feature-flagged sudo mode for 2fa setting

davidcornu opened the pull request:
RFC: #10653

The following PRs were extracted from this one and shipped separately:
- https://github.com/hackclub/hcb/pull/10780
- https://github.com/hackclub/hcb/pull/10826
- https://github.com/hackclub/hcb/pull/10903

## Changes

- Adds the model relationships described in https://github.com/hackclub/hcb/pull/10903#issue-3220305247 
  - Adds `logins` and `initial_login` relations to `UserSession`
  - Adds `initial_login` relation to `Login`
- Updates the logic in `Login` to only require a single factor in the case of a reauthentication
- Adds `UserSession#sudo_mode?` to determine whether sudo mode is enabled based on the creation times of its associated logins
- Adds `SudoModeHandler` which handles intercepting requests and rendering the reauthentication page
- Requires sudo mode if the 2fa setting is changed and the user has the feature flag enabled
- Extensive testing

<img width="375" height="667" alt="image" src="https://github.com/user-attachments/assets/43a03c4f-186a-451f-83a3-da838c4465c3" />

davidcornu committed:
Add `initial_login_id` to the `logins` table

davidcornu committed:
Add `UserSession#sudo_mode?`

davidcornu committed:
Add model relations

davidcornu committed:
Only require a single factor for reauthentications

davidcornu committed:
Add a very rough `enforce_sudo_mode` implementation

davidcornu committed:
Add `sudo_mode?` to `SessionsHelper`

davidcornu committed:
Remove memoization from `last_authenticated_at`

sampoder commented on app/controllers/users_controller.rb:
```diff
@@ -20,6 +20,7 @@ class UsersController < ApplicationController
                                                :complete_sms_auth_verification,
                                                :start_sms_auth_verification]
   before_action :set_shown_private_feature_previews, only: [:edit, :edit_featurepreviews, :edit_security, :edit_admin]
+  before_action :enforce_sudo_mode, only: [:update]
```
This is super clean - thank you for implementing it this way

sampoder commented on app/models/login.rb:
The changes to this model LGTM - only comment is I would have implemented `preceding_login` instead of inital

davidcornu replied:
That was my original plan but the approach I ended up landing on is to have everything point to the original login rather than have a chain in order to maintain the original login's lifetime (I felt it was simpler that way). Happy to switch it up if there's a compelling reason though.

sampoder commented on app/views/sudo_mode/reauthenticate.html.erb:
```diff
@@ -0,0 +1,21 @@
+
+<%= form_tag(request.path, method: request.request_method_symbol) do %>
+  <%
+    Rack::Utils
+      .parse_query(request.request_parameters.to_query)
```
I think we could make this a partial that we include in all the forms rendered from the `LoginsController` and then have some shared logic in `LoginsController#complete` to handle input from that.

sampoder reviewed (commented):
Model changes LGTM - as you mentioned in the video, I think the big thing will be reusing as much of the login controller code as possible. I'm happy to jump on a call to talk through that stuff at some point.

davidcornu committed:
Make rubocop happy

davidcornu committed:
Make `erb_lint` happy

davidcornu committed:
Move factor availability to `Login`

davidcornu committed:
Add `Login#available_factors`

davidcornu committed:
Get rid of `set_available_methods`

davidcornu committed:
Introduce `ProcessLoginService`

davidcornu committed:
Start adding test coverage

davidcornu committed:
Introduce `WebAuthnSupport`

davidcornu committed:
Flesh out webauthn tests

davidcornu commented:
Rebased on top of https://github.com/hackclub/hcb/pull/10826 to shrink the diff

davidcornu committed:
Use a local for `login`

davidcornu committed:
UI work

davidcornu committed:
Dedupe text field logic

davidcornu committed:
Fix tooltips on logo and mention

davidcornu committed:
Use `formnovalidate` on buttons that switch the method

davidcornu committed:
Extract logic into separate class

davidcornu committed:
Move factor preference into separate method

davidcornu committed:
Move params logic out of view

davidcornu committed:
Fix bad rename

davidcornu committed:
Many improvements

- Use a struct for sudo params
- Handle more authentication methods
- Break `call` up into separate methods

davidcornu committed:
Make prettier happy

davidcornu committed:
Merge `main` and fix conflicts

davidcornu committed:
Add test coverage for changes to `Login`

davidcornu committed:
Add test coverage for changes to `UserSession`

davidcornu committed:
Make extra sure we're only dealing with complete logins

davidcornu committed:
Remove unwanted `schema.rb` change

davidcornu committed:
Add test coverage for `SudoModeHandler`

davidcornu committed:
Raise if there isn't an initial login

davidcornu committed:
Add specific handling for repeated params

davidcornu committed:
Exclude backup codes

davidcornu committed:
Fix error case

davidcornu committed:
Avoid shadowing `login` variable

davidcornu committed:
Create a login in `SessionSupport#sign_in`

davidcornu committed:
Enforce sudo mode when 2fa settings are changed

davidcornu committed:
Make the return value of `enforce_sudo_mode` meaningful

davidcornu committed:
Put sudo mode behind `sudo_mode_2015_07_21` flag

davidcornu committed:
Move flipper logic to model

garyhtou commented on app/views/sudo_mode/reauthenticate.html.erb:
```diff
@@ -0,0 +1,95 @@
+        class: "btn mt-4"
+      ) %>
+
+  <% if additional_factors.present? %>
+    <p>Having problems?</p>
+    <ul>
+      <% additional_factors.each do |factor| %>
+        <li>
+          <%= button_tag(
+                type: "submit",
+                formnovalidate: true,
+                name: "_sudo[switch_method]",
+                value: factor.to_s,
+                class: "link-button"
+              ) do %>
+            <% case factor %>
+            <% when :email %>
+              Get a login code by email
+            <% when :sms %>
+              Get a login code by SMS
+            <% when :totp %>
+              Use a one-time password
+            <% when :webauthn %>
+              Use your security key
+            <% end %>
+          <% end %>
+        </li>
+      <% end %>
+    </ul>
+  <% end %>
```
Oh my gosh, this is sooo much cleaner than our current normal login UI

garyhtou commented on app/controllers/sudo_mode_handler.rb:
```diff
@@ -0,0 +1,218 @@
+  # - `:backup_code` is explicitly omitted as we don't want to use that for
+  #   reauthentication
+  FACTOR_PREFERENCES = {
+    totp: 1,
+    webauthn: 2,
```
Should we switch the order of `webauthn` and `totp`? I wonder if WebAuthn is more secure espeically since it supports hardware security keys and biometrics.

garyhtou commented on app/helpers/sessions_helper.rb:
```diff
@@ -158,4 +158,23 @@ def sign_out_of_all_sessions(user = current_user)
+  def enforce_sudo_mode
+    return true if sudo_mode?
+
+    SudoModeHandler.new(controller_instance: self).call
```
clever!

garyhtou reviewed (commented):
So far so good! I left a few comments and will come back to finish my review

garyhtou reviewed (approved):
Very nice! I'll need to spend some more time understanding the parameters, but i'm happy with where we're at right now!

Please go ahead and merge this before the sync!

garyhtou commented:
Oh yes, and a screenshot or two in the PR description would be nice

davidcornu committed:
Make WebAuthn the most favored authentication factor
